openforcefield/topology/molecule.py

```diff
@@ -1888,64 +1887,185 @@ def from_smiles(smiles,
         else:
             raise Exception(
                 'Invalid toolkit_registry passed to from_smiles. Expected ToolkitRegistry or ToolkitWrapper. Got  {}'
-                .format(type(toolkit_registry)))
+                    .format(type(toolkit_registry)))
 
-    def is_isomorphic(
-            self, other,
-            compare_atom_stereochemistry=True,
-            compare_bond_stereochemistry=True,
+    @staticmethod
+    def are_isomorphic(
+            mol1, mol2, return_atom_map=False,
+            aromatic_matching=True,
+            formal_charge_matching=True,
+            bond_order_matching=True,
+            atom_stereochemistry_matching=True,
+            bond_stereochemistry_matching=True,
     ):
         """
-        Determines whether the molecules are isomorphic by comparing their graphs.
+        Determines whether the two molecules are isomorphic by comparing their graphs.
```

Comment:
We should clarify to users what exactly is checked. For example, does this look at partial charges? It would be helpful to know all of the datapoints used to determine isomorphism.

Reply:
I think the new full docstring along with the kwargs should explain all node/edge attributes which can be checked. Let me know if this is clearer?

Reply:
Yes, the function is good as it stands, but I was just concerned as this function seems like a core piece of functionality that is really needed. I am interpreting this function as follows. Isomorphism is defined as whether it has the same atoms (nodes) and connectivity (edges). This satisfies the definition of isomorphism. Then, it goes and establishes that several auxiliary properties satisfy an equality test, and the function returns `False` if any of these are not equal, `True` otherwise.

Because of the way I see it working, I was concerned with what should be considered. Sitting and thinking more, I understand the spirit of the function and think it's nice for a subset of users. However, the way I'd probably use it, is set them all to `False`, then cherry pick the attributes I want to check manually, so I can figure out what is different. If the two mols are ordered the same, and this function returns `True` with all options as `False`, then I can just go down the list with an equality operator and check these individually.

Reply:
Yes, you are right the minimum requirement for the isomorphic check is the atomic numbers on the nodes and the connectivity (edges with no attributes) all the others are optional extras which can also be compared if users want to. I think your right that the majority of users will only use the minimum checking with all extras set to `False`, it might be worth at this point changing the default behaviour to `False` for easier default use, @j-wags would you agree?

Reply:
> I think your right that the majority of users will only use the minimum checking with all extras set to False,

@jthorton I don't think this is true. In the context of OFFTK, when most users want to know if two molecules are "identical", they mean "one molecule would have the same parameters assigned as another". My understanding of how this sort of equality could be checked is by 
1) comparing canonical SMILES, or 
2) having all the keyword arguments to `is_isomorphic` set to `True` except `atom_index` (and maybe `bond_order`, since the "same" molecule could have two valid kekule resonance structures, so we'd just check `is_aromatic`). 

> However, the way I'd probably use it, is set them all to False, then cherry pick the attributes I want to check manually, so I can figure out what is different

@trevorgokey This seems like an unusual use case -- When do we have a situation when a person has two molecules that differ only by (for example) one atom's element, and are looking for a function that tells them that? I'm imagining a case where a user compares two nearly-identical molecules -- one with a tetravalent N+ and the other with a tetravalent C in the same place -- and wants the OFFTK to tell them that these differ by element and formal charge. Maybe I'm not thinking of a more common use case, but that seems pretty far removed from OFFTK's core mission (parameterize molecules). People looking to do that sort of cheminformatics analysis should probably use RDKit. 

So, to summarize: 
* I think that uncolored graph comparison of molecules is an unusual thing to do outside of fairly involved cheminformatics
* OFFTK users are not looking to do fairly-involved cheminformatics (they can instead use RDKit)
* a good number of users will be unpleasantly surprised if our API used a different meaning of "isomorphism" than they had in mind, and people would shoot themselves in the foot by making the wrong assumption about the API.

> it might be worth at this point changing the default behaviour to False for easier default use, @j-wags would you agree?

@trevorgokey and I talked about this for some time yesterday, I do agree that "isomorphism" most directly implies "two graphs containing uncolored nodes connected by uncolored edges are superimposable" so we are technically using the word incorrectly. What we are really performing in the existing `Molecule.is_isomorphic' function is a _colored_ isomorphism check (where colors are element, bond order, etc). So, we are currently _technically_ using the word wrong, but to use it correctly would cause our users to shoot themselves in the foot

This is going to be an API breaking change either way (`is_isomorphic` is being deleted, and its replacement `is_isomorphic_with` uses different kwargs) , so we have some flexibility to start fresh here.

So this is a tricky call. I propose the following:

* We rename the current `are`/`is_isomorphic_with` to something more descriptive like `Molecule.map_molecule_graphs` or `compare_molecule_graphs`, which communicates that it will be more general
* Given that OFFTK is a parameterization engine foremost, we ensure that the `FrozenMolecule.__eq__` operator (used for `offmol1 == offmol2`) means "offmol1 would be assigned the same parameters as offmol2". In my mind this means "same canonicalized SMILES', but it may also correspond to a combination of graph comparison kwargs (though resonance structures/tautomers are tricky wrt. atom formal charge and bond orders) 
* We provide any number of "convenience methods" for molecule comparison, which basically are just pre-wrapped versions of `compare_molecular_graphs` with different keywords set.

@dmobley @jchodera This seems like a part of the API that's at high risk for foot-shooting, so your thoughts would be appreciated.

Reply:
Good point that was probably an oversight part and exact matches will be useful and important when it comes to deduplication of fragments in the bespoke workflow for example. 

I understand where you guys are coming from with the naming but this is still a type of isomorphic checking and in my mind, in this area, I think it would be rare for someone to only want to compare uncoloured graphs unless they were setting up some relative free energy calculation and needed a mapping between two topologies but this is beyond the scope of the toolkit. 

Following on from that  the return result (`True` or `False`) is expected with the naming `Molecule.is_isomorphic_with(other)` and `Molecule.are_isomorphic(mol1, mol2)` however, I don't feel it is clear with `Molecule.compare_molecule_graphs()`. Something like `Molecule.atoms_and_bonds_match(other)` is more descriptive and has an expected return type, but I would prefer the current naming.

Reply:
Posting a Slack DM discussion with some interesting API points: 

> Trevor Gokey  23:55
> Hey Jeff, I was playing around with the isomorph thing after reading your new comments. What do you think of this pseudocode? It envisions how I see using isomorphs, at least in my cases with dealing with the wild wild QCArchive. I see things this way because I am often trying to debug/categorize molecules that I know nothing about, and need to tease them apart. For me this works way better than a single function that just says yes/no isomorphic if I am e.g. trying to figure out why parameters are not the same between molecules.
```
def describe(mol1, mol2):
    exact = True
    #### check to see if heavy atom structure is the same to catch e.g. tautomers
    #### check formal charges ignoring hydrogen, this skips tautomer checks
    map = isomorphic(mol1, mol2, colors=None, ignore_H=True)
    if not map:
        return False
    print("Same heavy atom structure")
    equality_ops = [ eq.symbols, eq.aromaticity, eq.formal_charges ]
    for eq in equality_ops:
        if eq(mol1, mol2, map):
            print( eq, "True")
        else:
            exact = False
    ##### check for complete-atom isomorphs, with hydrogen
    ##### if different hydrogen and different formal charge, probably tautomer
    ##### if different hydrogen and same formal_charge, then different total charge (partial_charge)
    ##### need to establish same hydrogen before asking about stereochemistry, so check only if the molecule is completely isomorphic
    map = isomorphic(mol1, mol2, colors=None, ignore_H=False)
    if not map:
        print("Not same hydrogens: Tautomers?")
        return False
    equality_ops = [ eq.stereocenters, eq.parameter_labels, eq.formal_charges, eq.partial_charges ]
    for eq in equality_ops:
        if eq(mol1, mol2, map):
            print( eq, "True")
        else:
            exact = False
    if exact:
        print("Same exact molecule")
    return exact
```

> So you can see it is straightforward to check any other parameters or attributes in the molecules, as long as isomorphism is established. It is obviously very rough and maybe missing a few key points, but wanted to know your input before putting on the github discussion. I can see using colors as a shortcut to just check internally, making it easier for the user, e.g. isomorphic(mol1, mol2, colors=[eq.symbols, eq.aromaticity]) . Of course, here in my case by default colors=None and ignore_H=False .


> Jeffrey Wagner  12:44
> I have a lot of thoughts on this:
> * That's a beautiful API example, and I like the idea of using enums instead of a bunch of kwargs. 
> * Your use case is still weird, but I understand it better now. In some sense, I'd call what you're doing "dangerous", but I'm coming around to the idea that some fraction of users will also do that sort of analysis, no matter what we think. So if we can provide them clear, well-designed isomorphism tools, we'll enable more cool, reproducible science. And if we don't have to go too far out of the way to include that, it's a worthwhile goal.
> * I'm thinking that, if we three stakeholders have this many opinions on this API/functionality, then other toolkit users probably have a bunch of other opinions that I can't conceive of as well. The new isomorphism functionality isn't in the master branch yet, so we haven't broken the public API. We could keep all of the new code in the PR, but hide the parts under discussion as private functions (so are_isomorphic --> _are_isomorphic). This would start letting power users (like you and me) start banging on the underlying algorithms, while we continue to discuss the API. 


> Trevor Gokey  12:58
> As a small implementation detail, the enum you see is actually a list of functions, at least in the way I see it. Of course it could be implemented in various ways. I really think some nice public API to establish equality of our data is important. For example, FrozenMolecule.__eq__ could just be applying all of those `eq` functions.
> As for the dev versions: Yes, as we bang on this a bit, I definitely agree that we should have some "dev names" for these functions. Maybe not prepending "_" as it technically marks them private. Maybe a new namespace, e.g. Molecule.dev.are_isomorphic.
> In any case, a low level and high level API are pretty typical, so having both is a Good Thing(tm)

Reply:
I think I have a bit of a better understanding of this use case though I would still say its a rare one. I think a function like this constructed by a user could be achieved using the proposed `are_isomorphic` function with one additional kwarg of `include_hs=True` which should help you catch tautomers we could even expose a function to try and compare molecules and identify tautomers such as this `Molecule.are_tautomers(mol1, mol2)`. We would also need to change all `colour` matching options to `False` then you should be able to iterate over the matching where the enums represent kwarg dicts for example

```python
matching_options = [{'formal_charge_matching': True}, {'bond_order_matching': True}]
for kwargs in matching_options:
    if Molecule.are_isomorphic(mol1, mol2, **kwargs):
        # do somthing with the molecules
```

This would also mean that the default matching is basic graph matching (colourless edges and nodes with `atomic_number` attribute matching).